stan/math/opencl/kernel_generator/evaluate_into.hpp

```diff
@@ -0,0 +1,62 @@
+#ifndef STAN_MATH_OPENCL_KERNEL_GENERATOR_EVALUATE_INTO_HPP
+#define STAN_MATH_OPENCL_KERNEL_GENERATOR_EVALUATE_INTO_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/prim/scal/err/check_positive.hpp>
+#include <stan/math/opencl/kernel_generator/operation.hpp>
+#include <stan/math/opencl/kernel_generator/as_operation.hpp>
+#include <stan/math/opencl/kernel_generator/is_valid_expression.hpp>
+#include <cl.hpp>
+#include <string>
+#include <set>
+
+namespace stan {
+namespace math {
+
+template <typename Derived, typename ReturnScalar, typename... Args>
+template <typename T_lhs>
+void operation<Derived, ReturnScalar, Args...>::evaluate_into(
+    const T_lhs& lhs) const {
+  static_assert(
+      is_valid_expression<T_lhs>::value,
+      "operation::evaluate_into: left hand side is not a valid expression!");
+  using cache = operation<Derived, ReturnScalar, Args...>::cache<T_lhs>;
+  const auto& lhs_expression = as_operation(lhs);
+
+  int n_rows = derived().rows();
+  int n_cols = derived().cols();
+  const char* function = "evaluate_into";
+  check_positive(function, "number of rows", n_rows);
+  check_positive(function, "number of columns", n_cols);
+  check_size_match(function, "Rows of ", "*this", n_rows, "rows of ",
+                   "lhs_expression", lhs_expression.rows());
+  check_size_match(function, "Columns of ", "*this", n_cols, "columns of ",
+                   "lhs_expression", lhs_expression.cols());
+  try {
+    if (cache::kernel() == NULL) {
+      std::string src = get_kernel_source_for_evaluating_into(lhs);
+      auto opts = opencl_context.base_opts();
+      cache::kernel = opencl_kernels::compile_kernel(
```

Comment:
Could we make cache also hold the kernel string?

Reply:
That would be trivial to add. But why would we need it?

Reply:
This is mostly related to printing the kernels

Reply:
OK, I guess it could be sometimes useful for debugging.